rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,429 @@
+ - PIV provides additional useful functionality, including generating new certificates, setting a key's PIN and other management data
+ - PIV provides a useful standard for PIN and touch requirements, which will give us a universal UX across PIV devices
+ - new hardware keys are likely to implement a PIV module (Such as the [solo2](https://solokeys.com/blogs/news/update-on-our-new-and-upcoming-security-keys))
+
+If we want to expand support for non-PIV cryptographical devices in the future, we could add this functionality separately. TPM support in particular could be used to broaden hardware-backed key support to a wide array of modern computers, including all Windows 11 systems.
+
+#### Non-yubikey device compatibility
+
+Currently, Yubikey is one of the only PIV-compatible commercial hardware keys. As a result, current PIV implementations are specifically designed around Yubikey's implementation of PIV - the `libykcs11.so` module. While the majority of PIV is standardized, `libykcs11.so` has [some extensions](https://developers.yubico.com/PIV/Introduction/Yubico_extensions.html) and other idiosyncracies which may not be standard accross future PIV implemenations.
+
+Unfortunately, there is no common PIV library, so our best option is to use `piv-go` for a streamlined implemenation and prepare to adjust in the future as more PIV-compatible hardware keys are released. Possible adjustments include:
+ - using multiple PIV libraries to support custom PIV implementations
+ - switching to a PIV library which expressly supports all/more PIV implemenations
+ - working within a PIV library, through PRs or a Fork, to expand PIV support
+ - creating our own custom PIV library which we can add custom support into as needed
+
+Note: the adjustments above will largely be client-side and therefore should not pose any backwards compatibility concerns.
+
+### tsh PIV integration
+
+We'll add the `--piv` flag so that users can use `tsh login --piv` to login with their PIV card. The majority of the login flow is the same, except the key generation and certificate signing process.
+
+When `tsh login` would usually generate a new private key in memory, it will instead use [`piv.GenerateKey`](https://pkg.go.dev/github.com/go-piv/piv-go/piv#YubiKey.GenerateKey) to generate a new key pair directly on the PIV Authentication slot (`9a`). As usual, the certificate signing process will use the public key to generate and store certs in `~/.tsh/keys/proxy`.
+
+Neither the public or private key will be stored, as the private key can't be exported and we can later access the public key by checking the hardware key's [slot cert](https://pkg.go.dev/github.com/go-piv/piv-go/piv#YubiKey.Attest).
```

Comment:
https://docs.yubico.com/hardware/yubikey/yk-5/tech-manual/yk5-apps.html#slot-information

Reply:
To be honest, I'm not sure storing the public key is ever necessary. Instead we can just grab the public key from the private key. But I'll keep this in mind, at best it might improve performance slightly.